### PR TITLE
SDKS-1994_New_SecuredKey_Algorithm - For SDK 4.0 Release

### PR DESCRIPTION
**Pull request by george-bafaloukas-forgerock**
WIP updating the securedKey algorithm with newer suggested variant in a non-breaking way

# JIRA Ticket

Please, link jira ticket here.

# Description

Briefly describe the change and any information that would help speedup the review and testing process.

# Definition of Done Checklist:

- [ ] Acceptance criteria is met.
- [ ] All tasks listed in the user story have been completed.
- [ ] Coded to standards.
- [ ] Ensure backward compatibility.
- [ ] API reference docs is updated.
- [ ] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] Functional spec is written/updated.
- [ ] Example code snippets have been added.
- [ ] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

### CODE REVIEW

File: FRCore/FRCore/Keychain/SecuredKey.swift [hunk 1]
```diff
@@ -22,7 +22,7 @@ public struct SecuredKey {
     /// Public Key of SecuredKey
     fileprivate var publicKey: SecKey
     /// Algorithm to be used for encryption/decryption using SecuredKey
-    fileprivate let algorithm: SecKeyAlgorithm = .eciesEncryptionCofactorX963SHA256AESGCM
+    fileprivate let oldAlgorithm: SecKeyAlgorithm = .eciesEncryptionCofactorX963SHA256AESGCM//.eciesEncryptionCofactorVariableIVX963SHA256AESGCM
```
Comment (jeyanthanperiyasamy) [inline FRCore/FRCore/Keychain/SecuredKey.swift:25]: //.eciesEncryptionCofactorVariableIVX963SHA256AESGCM - can we remove the commented code ?

**Comment by jeyanthanperiyasamy** [comment]
i feel this need more testing to make sure we didnt break anything, see if we can cover any unit test for this added code.

**Review by jeyanthanperiyasamy — CHANGES_REQUESTED** [review]
i feel this need more testing to make sure we didnt break anything, see if we can cover any unit test for this added code.

File: FRCore/FRCore/Keychain/SecuredKey.swift [hunk 2]
```diff
@@ -180,17 +180,25 @@ public struct SecuredKey {
...
         }
         
         var error: Unmanaged<CFError>?
-        let decryptedData = SecKeyCreateDecryptedData(privateKey, algorithm, data as CFData, &error) as Data?
+        let decryptedData = SecKeyCreateDecryptedData(privateKey, secAlgorithm, data as CFData, &error) as Data?
         if let error = error {
-            Log.e("Failed to decrypt data: \(error)")
+            Log.e("Failed to decrypt data -  attempting Legacy Algorithm: \(error)")
+            var decryptError: Unmanaged<CFError>?
+            let decryptedData = SecKeyCreateDecryptedData(privateKey, oldAlgorithm, data as CFData, &decryptError) as Data?
```
Comment (jeyanthanperiyasamy) [inline FRCore/FRCore/Keychain/SecuredKey.swift:195]: why do we need to keep the old algorithm ? - i believe its for backward compatibility , mainly whoever already has the key encrypted using old algorithm still has to decrypt

File: FRCore/FRCore/Keychain/SecuredKey.swift [hunk 3]
```diff
@@ -161,15 +161,15 @@ public struct SecuredKey {
     
     /// Encrypts Data object using SecuredKey object
     /// - Parameter data: Encrypted Data object
-    public func encrypt(data: Data) -> Data? {
+    public func encrypt(data: Data, secAlgorithm: SecKeyAlgorithm = .eciesEncryptionCofactorVariableIVX963SHA256AESGCM) -> Data? {
         
-        guard SecKeyIsAlgorithmSupported(publicKey, .encrypt, algorithm) else {
-            Log.e("\(algorithm) is not supported on the device.")
+        guard SecKeyIsAlgorithmSupported(publicKey, .encrypt, secAlgorithm) else {
+            Log.e("\(oldAlgorithm) is not supported on the device.")
```
Comment (jeyanthanperiyasamy) [inline FRCore/FRCore/Keychain/SecuredKey.swift:167]: can you confirm is this suppose to be   - Log.e("\(secAlgorithm) is not supported on the device.") ?

File: FRCore/FRCore/Keychain/SecuredKey.swift [hunk 4]
```diff
@@ -180,17 +180,25 @@ public struct SecuredKey {
     
     /// Decrypts Data object using SecuredKey object
     /// - Parameter data: Decrypted Data object
-    public func decrypt(data: Data) -> Data? {
+    public func decrypt(data: Data, secAlgorithm: SecKeyAlgorithm = .eciesEncryptionCofactorVariableIVX963SHA256AESGCM) -> Data? {
         
-        guard SecKeyIsAlgorithmSupported(privateKey, .decrypt, algorithm) else {
-            Log.e("\(algorithm) is not supported on the device.")
+        guard SecKeyIsAlgorithmSupported(privateKey, .decrypt, secAlgorithm) else {
+            Log.e("\(oldAlgorithm) is not supported on the device.")
```
Comment (jeyanthanperiyasamy) [inline FRCore/FRCore/Keychain/SecuredKey.swift:186]: same as above can you confirm is this suppose to be  -  Log.e("\(secAlgorithm) is not supported on the device.") ?

**Review by jeyanthanperiyasamy — CHANGES_REQUESTED** [review]
minor comments

File: FRCore/FRCore/Keychain/SecuredKey.swift (hunk 3 above)
Comment (george-bafaloukas-forgerock) [reply FRCore/FRCore/Keychain/SecuredKey.swift:167]: good catch

File: FRCore/FRCore/Keychain/SecuredKey.swift (hunk 4 above)
Comment (george-bafaloukas-forgerock) [reply FRCore/FRCore/Keychain/SecuredKey.swift:186]: good catch

File: FRCore/FRCore/Keychain/SecuredKey.swift [hunk 5]
```diff
@@ -22,7 +22,7 @@ public struct SecuredKey {
     /// Public Key of SecuredKey
     fileprivate var publicKey: SecKey
     /// Algorithm to be used for encryption/decryption using SecuredKey
-    fileprivate let algorithm: SecKeyAlgorithm = .eciesEncryptionCofactorX963SHA256AESGCM
+    fileprivate let oldAlgorithm: SecKeyAlgorithm = .eciesEncryptionCofactorX963SHA256AESGCM
```
Comment (rodrigoareis) [inline FRCore/FRCore/Keychain/SecuredKey.swift:25]: Update copyright

File: FRCore/FRCoreTests/FRCore/Util/SecuredKeyTests.swift [hunk 6]
```diff
@@ -36,8 +36,7 @@ class SecuredKeyTests: FRBaseTestCase {
         
         let testString = "decrypted_text_for_testing".data(using: .utf8)!
         let encryptedUsingSecuredKey = key1?.encrypt(data: testString)
-        
-        let decryptedStringData = self.decrypt(data: encryptedUsingSecuredKey!, privateKey: privateKey, algorithm: .eciesEncryptionCofactorX963SHA256AESGCM)
+        let decryptedStringData = key1?.decrypt(data: encryptedUsingSecuredKey!)
```
Comment (rodrigoareis) [inline FRCore/FRCoreTests/FRCore/Util/SecuredKeyTests.swift:39]: Update copyright

**Review by spetrov — APPROVED** [review]
LGTM! 👍🏻

**Review by jeyanthanperiyasamy — APPROVED** [review]
LGTM

**Review by rodrigoareis — APPROVED** [review]
Looks good to me

**Pull request merged**